### ISSUE TITLE
Don't suggest `ref` for `let` bindings.

### ISSUE THREAD

**Issue by steveklabnik** [rust-lang/rust#40402]
**UPDATE:** There are some [initial mentoring instructions below](https://github.com/rust-lang/rust/issues/40402#issuecomment-288346360).

---

```rust
fn main() {
    let v = vec![String::from("oh no")];
    
    let e = v[0];
}
```

suggests

```text
error[E0507]: cannot move out of indexed content
 --> <anon>:4:13
  |
4 |     let e = v[0];
  |         -   ^^^^ cannot move out of indexed content
  |         |
  |         hint: to prevent move, use `ref e` or `ref mut e`
```

`let ref e` is not idiomatic here, so this is misleading (and confusing, frankly)

cc @eevee

**Comment by nikomatsakis** [comment]
@steveklabnik 

First, would you prefer to just stop hinting about `ref` in **any** let pattern, or just if you write `let e = ...`? 

Second, do you think we should hint something like "consider `let e = &v[0]`"? Perhaps only if the RHS is an lvalue? (Well, I guess this error will always result from an lvalue anyway.)

**Comment by nikomatsakis** [comment]
Fixing this requires doing a bit of plumbing. I don't have time to get into every detail so let me just jot down some notes about how the code works today (I was just reading it).

- The message is generated by the borrow-checker. It finds out about the move due to a call to `consume_pat()`, a callback invoked by the `ExprUseVisitor` (EUV) for moves from patterns. This callback does not currently distinguish moves that occur in `let` patterns from those that occur in `match` patterns, so the borrow-checker can't easily tell the difference. 
    - We could fix this by extending the EUV interface, but we could also walk up the HIR map (in `rustc::hir::map`) from the pattern to find out the enclosing context. The latter might be better, since we could distinguish e.g. `let x = ...` from `let Foo { x } = ...`, and also gain access to the RHS of the `let`.
- From the callback, the borrow checker gathers up moves in `librustc_borrowck/borrowck/gather_loans/gather_moves.rs`, and specifically the function `gather_move_from_pat()`.
    - This function creates a `GatherMoveInfo` struct whose field `span_path_opt` is `Some(_)`
- We then later report the errors. The code for this is found in `src/librustc_borrowck/borrowck/gather_loans/move_errors.rs`, and specifically the `report_move_errors()` function.
    - This will group together errors by location, and collect the `span_path_opt` fields into a vector. It then iterates over this vector (`error.move_to_places`) (which is empty if all the `span_path_opt` fields were `None`) to issue the `ref` notes.

So the most *direct* fix would be to notice that the pattern is in a let and set the field to `None`. But perhaps a better fix would be to detect cases where we can suggest writing `&v[0]` and do *that* instead. I'm not entirely sure what's the most elegant way to modify things here. I imagine we need to extend `MoveSpanAndPath` to contain the `NodeId` of the pattern. We could even replace the struct entirely, since the other info can be derived from the `NodeId`: we can use `bccx.tcx.hir.find(id)` and check that the result is the `Node::NodePat` variant (it must be) to extract the other information. Regardless, we can walk up the HIR map to find if the enclosing context for this pattern is a `let`, and also extract the initializer (the RHS of the `let`). 

Anyway, that's a bit dense, I can try to expand some in a follow-up comments later on.

**Comment by gaurikholkar** [comment]
Hi. Will take this up.

**Comment by steveklabnik** [comment]
> First, would you prefer to just stop hinting about ref in any let pattern, or just if you write let e = ...? (e.g., let Foo { ref e, ref g } = x is something I at least do fairly regularly...)

I am mostly thinking about the former, that is, the latter is idiomatic use of `ref`.

> Second, for the simple case, do you think we should hint something like "consider let e = &v[0]"? Perhaps only if the RHS is an lvalue? (Well, I guess this error will always result from an lvalue anyway.)

That would be great.

**Comment by cynicaldevil** [comment]
@nikomatsakis I tried coding up the `get_pattern_source` method you showed, and after I some corrections, I have this now: [https://gist.github.com/cynicaldevil/4f5a2e8b96bb0d47d89ecb7626da5ccd](https://gist.github.com/cynicaldevil/4f5a2e8b96bb0d47d89ecb7626da5ccd)

I have worked with Rust before, but I don't have much experience with lifetimes, and therefore I'm getting stuck on line 19, where the compiler throws an error saying that `local.clone().unwrap()` does not live long enough. I know that the lifetime of the value must be equal to the lifetime of the method's body, but I can't find a way to make it so. I did everything I could (adding/removing refs, cloning values), but no luck. Could you help me out?

**Comment by arielb1** [comment]
@cynicaldevil 

There's no point in cloning HIR nodes. You can always take a reference to them that has lifetime `'hir`.

**Comment by cynicaldevil** [comment]
@arielb1 Thanks, I solved it at last. Turns out, the whole problem was unrelated to lifetimes, I was actually having trouble extracting a value from a smart pointer, and was using the `unwrap` method the whole time, which moved the value into the match arm. I didn't know that the pointer implemented a `deref` too, which solved the issue for me.

**Comment by nikomatsakis** [comment]
@cynicaldevil sorry I've been unresponsive; been busy traveling around, hopefully better now. =) Everything working out ok?

**Comment by cynicaldevil** [comment]
@nikomatsakis no worries :) As you can see, I'm blocked because of #41169 now, so I don't know how to proceed further.

**Comment by nikomatsakis** [comment]
@cynicaldevil I see -- I can leave some comments on that issue to suggest how to resolve it. However, I also think that @gaurikholkar is still actively working on this as well... seems like a shame to duplicate effort. Not sure how far they have gotten.

**Comment by gaurikholkar** [comment]
Right now, I have written the ` get_pattern_source` function, [get_pattern_source](https://github.com/gaurikholkar/rust/commit/ff6c1d433cafa09ffa032c24dcc7b910cb5a6006) and for
```
fn main() {
    let v = vec![String::from("oh no")];
    
    let e = v[0];
}
```
it suggests

```
error[E0507]: cannot move out of indexed content
 --> example.rs:4:13
  |
4 |     let e = v[0];
  |             ^^^^ cannot move out of indexed content

error: aborting due to previous error
```

while for 
```
fn main() {
    let x = vec![(String::new(), String::new())];
    let (a, b) = x[0];
}
```
it suggests
```
error[E0507]: cannot move out of indexed content
 --> example1.rs:3:18
  |
3 |     let (a, b) = x[0];
  |          -  -    ^^^^ cannot move out of indexed content
  |          |  |
  |          |  ...and here (use `ref b` or `ref mut b`)
  |          hint: to prevent move, use `ref a` or `ref mut a`

error: aborting due to previous error
```
as we are only disabling the ref hint if let is the immediate parent node [here](https://github.com/gaurikholkar/rust/commit/39d01f7a56abe0f2bd7e0b19635dbc81062efc30)

I have written UI tests as follows
[ui tests](https://github.com/gaurikholkar/rust/commit/84e8b3440c194ccdf9ef37b0416b9adb79488d58)

**Comment by gaurikholkar** [comment]
#41564

**Comment by Mark-Simulacrum** [comment]
This was fixed in #41564 and #41640. Thanks @gaurikholkar!
